### PR TITLE
Don't duplicate region names for late-bound regions in print of Binder

### LINKED ISSUE #101280: Stand-in lifetime names for printing HRTs doesn’t avoid name collisions.
E.g.
```rs
use std::cell::Cell;

fn f<'a>(f: fn(Cell<(&'a i32, &i32)>)) -> for<'l> fn(Cell<(&'l i32, &'l i32)>) {
    f
}
```
chooses the lifetime name “`'r`” for the fn-pointer type of the argument `f` and prints
```
error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
3 | fn f<'a>(f: fn(Cell<(&'a i32, &i32)>)) -> for<'l> fn(Cell<(&'l i32, &'l i32)>) {
  |                                           ------------------------------------ expected `for<'l> fn(Cell<(&'l i32, &'l i32)>)` because of return type
4 |     f
  |     ^ one type is more general than the other
  |
  = note: expected fn pointer `for<'l> fn(Cell<(&'l i32, &'l i32)>)`
             found fn pointer `for<'r> fn(Cell<(&'a i32, &'r i32)>)`

For more information about this error, try `rustc --explain E0308`.
```
---
However, rewriting the code into
```rs
use std::cell::Cell;
type Ty = for<'r> fn(Cell<(&'r i32, &'r i32)>);
fn f<'r>(f: fn(Cell<(&'r i32, &i32)>)) -> Ty {
    f
}
```
does ***not*** currently make the compiler reconsider the choice of “`'r`” in the lifetime name, resulting in an unhelpful error message
```
error[E0308]: mismatched types
 --> src/lib.rs:4:5
  |
3 | fn f<'r>(f: fn(Cell<(&'r i32, &i32)>)) -> Ty {
  |                                           -- expected `for<'r> fn(Cell<(&'r i32, &'r i32)>)` because of return type
4 |     f
  |     ^ one type is more general than the other
  |
  = note: expected fn pointer `for<'r> fn(Cell<(&'r i32, &'r i32)>)`
             found fn pointer `for<'r> fn(Cell<(&'r i32, &'r i32)>)`
```

### PR DESCRIPTION
**Pull request by b-naber**
Fixes https://github.com/rust-lang/rust/issues/101280

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -103,9 +103,9 @@ dependencies = [
```
Comment (lcnr) [inline Cargo.lock:103]: these changes look unintended?

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 2]
```diff
@@ -2174,13 +2184,23 @@ impl<'tcx> FmtPrinter<'_, 'tcx> {
 
         define_scoped_cx!(self);
 
+        let possible_names = vec![Symbol::intern("'t"), Symbol::intern("'s"), Symbol::intern("'r")];
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/print/pretty.rs:2187]: maybe go from `'a` to `'z` before starting with `'t<n>` :thinking: so

```suggestion
 let possible_names: Vec<_> = ('a'..'z').map(|s| Symbol::intern(&format!("'{s}"))).collect();
````

File: compiler/rustc_middle/src/ty/sty.rs [hunk 3]
```diff
@@ -1505,6 +1516,23 @@ impl<'tcx> Region<'tcx> {
         *self.0.0
     }
 
+    pub fn get_name(self) -> Option<Symbol> {
+        if self.has_name() {
+            let name = match *self {
+                ty::ReEarlyBound(ebr) => Some(ebr.name),
+                ty::ReLateBound(_, br) => br.kind.get_name(),
+                ty::ReFree(fr) => fr.bound_region.get_name(),
+                ty::ReStatic => Some(Symbol::intern("static")),
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/sty.rs:1525]:
```suggestion
                ty::ReStatic => Some(Symbol::intern("'static")),
```
i think?

Comment (lcnr) [reply compiler/rustc_middle/src/ty/sty.rs:1525]: though probably add that as a predefined symbol to `sym`.

**Review by lcnr — COMMENTED** [review]
some nits, then r=me

File: compiler/rustc_middle/src/ty/sty.rs (hunk 3 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/ty/sty.rs:1525]:
```suggestion
                ty::ReStatic => Some(sym::StaticLifetime),
```

**Comment by matthiaskrgr** [comment]
@bors r- need test update it seems
https://github.com/rust-lang/rust/actions/runs/3108219234/jobs/5037174841#step:25:1567

**Comment by lcnr** [comment]
yeah, have to rebase and rebless ui-tests as a new affected test was added since you've forked master afaict

**Pull request merged**